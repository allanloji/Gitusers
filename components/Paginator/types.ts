type PaginatorProps = {
  currentPage: number;
  perPage: number;
  totalResults: number;
  setPage: (page: number) => void;
  prevPage: () => void;
  nextPage: () => void;
};

export type { PaginatorProps };
