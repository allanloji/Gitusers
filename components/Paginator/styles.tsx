import styled from '@emotion/styled';

const Container = styled.div`
  display: flex;
  align-items: center;
  margin: 2rem 0;
  justify-content: center;
  width: 100%;
`;

const Button = styled.button`
  width: 40px;
  height: 40px;
  border-radius: 20px;
  margin: 0 0.2rem;
  color: white;
  border: none;
  background-color: rgb(43, 43, 43);

  cursor: pointer;

  :disabled {
    background-color: rgb(90, 90, 90);
    color: rgb(150, 150, 150);
    cursor: not-allowed;
  }
`;
const PageText = styled.p`
  color: white;
  margin: 0 1rem;
`;
export { Container, Button, PageText };
