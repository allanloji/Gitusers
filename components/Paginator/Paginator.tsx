import React from 'react';

import { Button, Container, PageText } from './styles';
import { PaginatorProps } from './types';

/**
 * Paginator component
 */
function Paginator({
  currentPage,
  nextPage,
  prevPage,
  setPage,
  perPage,
  totalResults,
  isLoading,
}: PaginatorProps) {
  // Github API limits to the first 1000  results
  const maxAvailableResults = Math.min(totalResults, 1000);
  const totalPages = Math.ceil(maxAvailableResults / perPage);

  const isFirstPage = currentPage === 1 || isLoading;
  const isLastPage = currentPage === totalPages || isLoading;

  return (
    <Container>
      <Button
        type='button'
        onClick={() => setPage(1)}
        disabled={isFirstPage}
        data-testid='firstPage'
      >
        {'<<'}
      </Button>
      <Button type='button' onClick={prevPage} disabled={isFirstPage} data-testid='prevPage'>
        {'<'}
      </Button>
      <PageText>
        {currentPage} of {totalPages}
      </PageText>
      <Button type='button' onClick={nextPage} disabled={isLastPage} data-testid='nextPage'>
        {'>'}
      </Button>
      <Button
        type='button'
        onClick={() => setPage(totalPages)}
        disabled={isLastPage}
        data-testid='lastPage'
      >
        {'>>'}
      </Button>
    </Container>
  );
}

export default Paginator;
