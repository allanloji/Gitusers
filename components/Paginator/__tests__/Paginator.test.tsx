import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Paginator from '..';

describe('Paginator', () => {
  test('Should render', () => {
    const prevPage = jest.fn();
    const nextPage = jest.fn();
    const setPage = jest.fn();

    const { container } = render(
      <Paginator
        currentPage={1}
        perPage={9}
        prevPage={prevPage}
        nextPage={nextPage}
        setPage={setPage}
        totalResults={1000}
        isLoading={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('Should render blocked buttons when is loading', () => {
    const prevPage = jest.fn();
    const nextPage = jest.fn();
    const setPage = jest.fn();

    const { container } = render(
      <Paginator
        currentPage={1}
        perPage={9}
        prevPage={prevPage}
        nextPage={nextPage}
        setPage={setPage}
        totalResults={1000}
        isLoading
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('Should change pages', () => {
    const prevPage = jest.fn();
    const nextPage = jest.fn();
    const setPage = jest.fn();

    const { getByTestId } = render(
      <Paginator
        currentPage={5}
        perPage={9}
        prevPage={prevPage}
        nextPage={nextPage}
        setPage={setPage}
        totalResults={1000}
        isLoading={false}
      />,
    );

    const prevPageButton = getByTestId('prevPage');
    userEvent.click(prevPageButton);
    expect(prevPage).toHaveBeenCalled();

    const nextPageButton = getByTestId('nextPage');
    userEvent.click(nextPageButton);
    expect(nextPage).toHaveBeenCalled();

    const firstPageButton = getByTestId('firstPage');
    userEvent.click(firstPageButton);
    expect(setPage).toHaveBeenCalledWith(1);

    const lastPageButton = getByTestId('lastPage');
    userEvent.click(lastPageButton);
    expect(setPage).toHaveBeenCalledWith(112);
  });
});
