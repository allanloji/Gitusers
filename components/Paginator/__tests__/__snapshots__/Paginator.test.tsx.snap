// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator Should render 1`] = `
<div>
  <div
    class="css-1mkvne5"
  >
    <button
      class="css-q1rq5h"
      data-testid="firstPage"
      disabled=""
      type="button"
    >
      &lt;&lt;
    </button>
    <button
      class="css-q1rq5h"
      data-testid="prevPage"
      disabled=""
      type="button"
    >
      &lt;
    </button>
    <p
      class="css-v8p8i1"
    >
      1
       of 
      112
    </p>
    <button
      class="css-q1rq5h"
      data-testid="nextPage"
      type="button"
    >
      &gt;
    </button>
    <button
      class="css-q1rq5h"
      data-testid="lastPage"
      type="button"
    >
      &gt;&gt;
    </button>
  </div>
</div>
`;

exports[`Paginator Should render blocked buttons when is loading 1`] = `
<div>
  <div
    class="css-1mkvne5"
  >
    <button
      class="css-q1rq5h"
      data-testid="firstPage"
      disabled=""
      type="button"
    >
      &lt;&lt;
    </button>
    <button
      class="css-q1rq5h"
      data-testid="prevPage"
      disabled=""
      type="button"
    >
      &lt;
    </button>
    <p
      class="css-v8p8i1"
    >
      1
       of 
      112
    </p>
    <button
      class="css-q1rq5h"
      data-testid="nextPage"
      disabled=""
      type="button"
    >
      &gt;
    </button>
    <button
      class="css-q1rq5h"
      data-testid="lastPage"
      disabled=""
      type="button"
    >
      &gt;&gt;
    </button>
  </div>
</div>
`;
